File: doc/update_execution_example.py
#!/usr/bin/env python
# -*- coding: utf-8; -*-　

import subprocess
from select import select
from sys import argv,stdout
import os
import os.path
import shutil
import re
import json

DB_DIRECTORY = "/tmp/groonga-databases"
DEFAULT_DB_NAME = "tutorial.db"

shutil.rmtree(DB_DIRECTORY, ignore_errors=True)
os.makedirs(DB_DIRECTORY)

groonga_process = None
current_log = None
output_log = False
def close_groonga():
  global groonga_process
  global current_log
  global output_log
  if groonga_process:
    groonga_process.stdin.close()
    groonga_process.stdout.close()
    groonga_process = None
    print '###<<< database: close'
  if current_log:
    current_log.close()
    current_log = None
  output_log = False

current_db_path = None
current_log_path = None
def reconnect(name):
  global groonga_process
  global current_log
  global current_db_path
  global current_log_path
  close_groonga()
  current_db_path = os.path.join(DB_DIRECTORY, name)
  current_log_path = os.path.join(DB_DIRECTORY, name + ".log")
  if os.path.exists(current_db_path):
    groonga_process = subprocess.Popen(["groonga",
                                        "--log-path", current_log_path,
                                        current_db_path],
                                       stdin=subprocess.PIPE,
                                       stdout=subprocess.PIPE)
  else:
    groonga_process = subprocess.Popen(["groonga",
                                        "--log-path", current_log_path,
                                        "-n",
                                        current_db_path],
                                       stdin=subprocess.PIPE,
                                       stdout=subprocess.PIPE)
  groonga_process.stdin.write("status\n")
  groonga_process.stdin.flush()
  groonga_process.stdout.readline()
  current_log = open(current_log_path)
  current_log.read()
  print '###>>> database: open <%s>' % current_db_path

def expand_command_line(command_line):
  return command_line.replace('${DB_PATH}', current_db_path)

fout = None

def normalize_file_name(file_name):
  return re.sub('^.*?/lib/', 'lib/', file_name)

def normalize_output(output):
  status = output[0]
  if status:
    normalized_start_time = 1337566253.89858
    normalized_elapsed_time = 0.000355720520019531
    status[1] = normalized_start_time
    status[2] = normalized_elapsed_time
    return_code = status[0]
    if return_code != 0:
      backtraces = status[4]
      if backtraces:
        for backtrace in backtraces:
          file_name = backtrace[1]
          backtrace[1] = normalize_file_name(file_name)
  return output

def remove_continuous_marks(command):
  return re.sub('\\\\\n', '', command)

def prepend_prefix(prefix, command):
  beginning_of_line = re.compile('^', re.M)
  return beginning_of_line.sub(prefix, command)

def execmd(command, fout):
  stdout.write(command + "\n")
  stdout.flush()
  groonga_process.stdin.write(remove_continuous_marks(command) + "\n")
  groonga_process.stdin.flush()
  is_command = re.match("[a-z/]", command)
  is_load_command = re.match("load ", command)
  is_console = not re.match("/", command)
  if fout:
    if is_console:
      prefix = "  "
    else:
      prefix = "  % curl http://localhost:10041"
    formatted_command_line = prepend_prefix(prefix, command) + "\n"
    fout.write(formatted_command_line)
  is_load_data_end = re.match("^\]", command)
  if is_load_command:
    return
  if not is_command and not is_load_data_end:
    return
  output_buffer = ""
  first_timeout = 1
  timeout = first_timeout
  while True:
    out = select([groonga_process.stdout], [], [], timeout)
    timeout = rest_timeout
    if len(out[0]):
      char = groonga_process.stdout.read(1)
      if char is None or char == '':
        stdout.write(output_buffer)
        if fout:
          fout.write(output_buffer)
        break
      else:
        output_buffer += char
        if char == '\n':
          if command == 'dump':
            formatted_output = output_buffer
            stdout.write(output_buffer)
          else:
            parsed_output = json.loads(output_buffer)
            normalized_output = normalize_output(parsed_output)
            if len(output_buffer) < 80:
              formatted_output = json.dumps(normalized_output,
                                            ensure_ascii=False)
            else:
              formatted_output = json.dumps(normalized_output,
                                            indent=2,
                                            ensure_ascii=False)
              formatted_output += "\n"
              formatted_output = formatted_output.encode("utf-8")
            stdout.write(formatted_output)
            stdout.write("\n")
          if fout:
            if is_console:
              prefix = "  # "
            else:
              prefix = "  "
            first_lines_re = re.compile("^", re.M)
            fout.write(first_lines_re.sub(prefix, formatted_output.strip()))
            fout.write("\n")
          if current_log:
            log = current_log.read().strip()
            if len(log) > 0:
              prefix = "  # log: "
              first_lines_re = re.compile("^", re.M)
              formatted_log = first_lines_re.sub(prefix, log)
              stdout.write(first_lines_re.sub(prefix, log))
              stdout.write("\n")
              if output_log and fout:
                fout.write(formatted_log)
                fout.write("\n")
          output_buffer = ""
    else:
      stdout.flush()
      break

processed_files = []
def readfile(fname, outflag):
  if fname in processed_files:
    print "skipped processed file: %s" % fname
    return
  if outflag > 32:
    print "!!!! INCLUDE DEPTH OVER !!!!"
    raise
  processed_files.append(fname)

  b = fname.rfind('/')
  if b < 0:
    rootdir = './'
  else:
    rootdir = fname[0:b+1]

  fi = open(fname, 'r')
  dat = fi.read().split("\n")
  fi.close()

  line = 0;
  while len(dat):
    cmd = dat.pop(0)
    if cmd.startswith('.. groonga-command'):
      print '### command start'
      fout = None
      while len(dat):
        cmd = dat.pop(0)
        if cmd.startswith('.. database:'):
          database_name = cmd[cmd.index(":")+1:].strip()
          reconnect(database_name)
        elif cmd.startswith('.. log:'):
          global output_log
          log_value = cmd[cmd.index(":")+1:].strip()
          output_log = log_value == "true"
        elif cmd.startswith('.. include:: '):
          a = rootdir + cmd[13:]
          dir_name = os.path.dirname(a)
          if not os.path.exists(dir_name):
            os.makedirs(dir_name)
          fout = open(a, 'w')
          print '### write start : ' + a
          fout.write("Execution example::\n\n")
        elif cmd.startswith('.. % '):
          command_line = cmd[5:]
          if fout:
            fout.write("  % " + command_line + "\n")
          print command_line
          expanded_command_line = expand_command_line(command_line)
          command_output = subprocess.check_output(expanded_command_line,
                                                   shell=True)
          if fout:
            first_lines_re = re.compile("^", re.M)
            fout.write(first_lines_re.sub("  ", command_output.strip()))
            fout.write("\n")
          else:
            print command_output
        elif cmd.startswith('.. .. '):
          command_line = cmd[6:]
          if fout:
            fout.write("  " + command_line + "\n")
          print command_line
        elif cmd.startswith('..'):
          if cmd.replace(' ', '').replace("\t", '') == '..':
            while len(dat):
              if dat[0] == '' or (dat[0][0] != ' ' and dat[0][0] != '	'):
                break
              execmd(dat.pop(0), fout)
          else:
            cmd = cmd[3:]
            while cmd.endswith('\\'):
              cmd += '\n' + re.sub('^\.\. ', '', dat.pop(0))
            execmd(cmd, fout)
        else:
          print '### command end'
          if fout:
            fout.close()
          break
    elif cmd.startswith('.. groonga-include : '):
      a = rootdir + cmd[21:]
      print '###>>> include : ' + a
      readfile(a, outflag + 1)
      print '###<<< include end'

entry_point = "source/"
rest_timeout = 0.1
if len(argv) == 2:
  entry_point = argv[1]
elif len(argv) == 3:
  entry_point = argv[1]
  rest_timeout = float(argv[2])
if os.path.isfile(entry_point):
  readfile(entry_point, 0)
else:
  for root, dirs, files in os.walk(entry_point):
    for fname in files:
      if fname.lower().endswith('.rst'):
        b = os.path.join(root, fname)
        print "===" + b
        readfile(b, 0)

if fout:
  fout.close()
close_groonga()

File: doc/update-execution-example.rb
#!/usr/bin/env ruby
#
# Copyright (C) 2023-2024  Sutou Kouhei <kou@clear-code.com>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

require "fileutils"
require "find"
require "json"

class Updator
  def initialize(base_dir)
    @base_dir = base_dir
    @verbose = (ENV["VERBOSE"] == "true")
  end

  def update(source)
    FileUtils.rm_rf(@base_dir)
    FileUtils.mkdir_p(@base_dir)

    @processed_files = {}
    @current_db = nil
    @output_log = false
    Find.find(source) do |path|
      update_file(path)
    end
  end

  private
  def puts(*args)
    return unless @verbose
    $stderr.puts(*args)
  end

  def expand_command_line(command_line)
    command_line.gsub(/\${DB_PATH}/) do
      current_db_path
    end
  end

  def current_db_path
    File.join(@base_dir, @current_db)
  end

  def current_log_path
    "#{current_db_path}.log"
  end

  def run_groonga
    command_line = [
      "groonga",
      "--log-path", current_log_path,
    ]
    command_line << "-n" unless File.exist?(current_db_path)
    command_line << current_db_path
    FileUtils.touch(current_log_path)
    IO.pipe do |in_read, in_write|
      IO.pipe do |out_read, out_write|
        pid = spawn(*command_line, in: in_read, out: out_write)
        in_read.close
        out_write.close
        begin
          File.open(current_log_path) do |log|
            # Ensure initializing
            in_write.puts("status")
            in_write.flush
            out_read.gets

            log.seek(0, IO::SEEK_END)
            @current_log = log
            yield(in_write, out_read)
          end
        ensure
          @current_log = nil
          in_write.close unless in_write.closed?
          out_read.close unless out_read.closed?
          Process.waitpid(pid)
        end
      end
    end
  end

  def read_output(output)
    data = ""
    timeout = 10
    while IO.select([output], nil, nil, timeout)
      break if output.eof?
      data << output.readpartial(4096)
      timeout = 0
    end
    data
  end

  def parse_result(command, result)
    begin
      JSON.parse(result)
    rescue JSON::ParserError
      puts("Failed to parse:")
      puts("path: #{@current_path}")
      puts("command:")
      puts(command)
      puts("result:")
      puts(result)
      raise
    end
  end

  def normalize_error_message(message)
    message.gsub(current_db_path, "${DB_PATH}")
  end

  def normalize_file_name(file_name)
    file_name = file_name.gsub(/\A\/.*\/\.\.\/plugins\//, "lib/groonga/plugins/")
    file_name = file_name.gsub(/\A.*?\/lib\//, "lib/")
    file_name
  end

  def normalize_result(command, result)
    header = result[0]
    if header
      normalized_start_time = 1337566253.89858
      normalized_elapsed_time = 0.000355720520019531
      header[1] = normalized_start_time
      header[2] = normalized_elapsed_time
      header[3] = normalize_error_message(header[3]) if header[3]
      return_code = header[0]
      if return_code.zero?
        if command.start_with?("status") or command.start_with?("/d/status")
          status = result[1]
          status["alloc_count"] = 29 if status.key?("alloc_count")
          status["starttime"] = 1696558618 if status.key?("starttime")
          status["start_time"] = 1696558618 if status.key?("start_time")
          status["version"] = "2.9.1" if status.key?("version")
          apache_arrow = status["apache_arrow"]
          if apache_arrow
            if apache_arrow.key?("version_major")
              apache_arrow["version_major"] = 2
            end
            if apache_arrow.key?("version_minor")
              apache_arrow["version_minor"] = 9
            end
            if apache_arrow.key?("version_patch")
              apache_arrow["version_patch"] = 1
            end
            if apache_arrow.key?("version")
              apache_arrow["version"] = "2.9.1"
            end
          end
          status["memory_map_size"] = 2929 if status.key?("memory_map_size")
          status["os"] = "Linux" if status.key?("os")
          status["cpu"] = "x86_64" if status.key?("cpu")
        end
      else
        backtraces = header[4]
        if backtraces
          backtraces.each do |backtrace|
            file_name = backtrace[1]
            backtrace[1] = normalize_file_name(file_name)
            # line
            backtrace[2] = 2929 if backtrace[2]
          end
        end
      end
    end
    result
  end

  def normalize_formatted_result(formatted_result)
    formatted_result.gsub(current_db_path, "${DB_PATH}")
  end

  def normalize_log(log)
    log.gsub(/^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d+/,
             "2023-10-05 17:26:13.890356")
  end

  def detect_markup(path)
    File.extname(path)[1..-1].to_sym
  end

  def execute_command(input, output, command, current_output_path, output_log)
    markup = detect_markup(current_output_path)
    input.puts(command)
    input.flush
    is_command = /\A[a-z\/]/.match?(command)
    is_load_command = command.start_with?("load ")
    is_path_style = command.start_with?("/")
    puts(command)
    result = read_output(output)
    if command.start_with?("dump")
      formatted_result = result
    else
      parsed_result = parse_result(command, result)
      normalized_result = normalize_result(command, parsed_result)
      formatted_result = JSON.generate(normalized_result)
      if formatted_result.bytesize > 79
        formatted_result = JSON.pretty_generate(normalized_result)
      end
      formatted_result = normalize_formatted_result(formatted_result)
    end
    puts(formatted_result)
    if current_output_path
      File.open(current_output_path, "a") do |o|
        command_prefix = +""
        command_prefix << "  " if markup == :rst
        command_prefix << "$ curl http://localhost:10041" if is_path_style
        o.puts(command.gsub(/^/, command_prefix))
        output_prefix = +""
        output_prefix << "  " if markup == :rst
        output_prefix << "# " unless is_path_style
        o.puts(formatted_result.gsub(/^/, output_prefix))

        if output_log
          log = normalize_log(read_output(@current_log).strip)
          unless log.empty?
            log_prefix = "# log: "
            puts(log.gsub(/^/, log_prefix))
            o.puts(log.gsub(/^/, "  #{log_prefix}"))
          end
        end
      end
    end
  end

  def update_file(path)
    if @processed_files.key?(path)
      puts("Skipped processed file: #{path}")
      return
    end
    case File.extname(path)
    when ".rst"
      @processed_files[path] = true
      @current_path = path
      update_rst(path)
    when ".md"
      @processed_files[path] = true
      @current_path = path
      update_md(path)
    end
  end

  def update_rst(path)
    groonga_command_block = ""
    in_groonga_command = false
    File.read(path).each_line do |line|
      if in_groonga_command
        if line.start_with?("..")
          groonga_command_block << line.gsub(/\A\.\. ?/, "")
        else
          in_groonga_command = false
          process_groonga_command(groonga_command_block)
          groonga_command_block.clear
        end
      else
        case line
        when /\A\.\. groonga-command/
          in_groonga_command = true
        when /\A\.\. groonga-include\s*:/
          include_path = line.split(":", 2)[1].strip
          update_file(File.join(File.dirname(@current_path), include_path))
        end
      end
    end
    unless groonga_command_block.empty?
      process_groonga_command(groonga_command_block)
    end
  end

  def update_md(path)
    groonga_command_block = ""
    in_groonga_command = false
    in_include = false
    File.read(path).each_line do |line|
      if in_groonga_command
        if in_include
          if line.chomp == "```"
            in_groonga_command = false
            in_include = false
            process_groonga_command(groonga_command_block)
            groonga_command_block.clear
          else
            groonga_command_block << line
          end
        else
          case line.chomp
          when /\A```{include} /
            in_include = true
            groonga_command_block << line.gsub(/\A```{include} /, "include::")
          when /\A<!-- database: /
            @current_db = line.split(":", 2)[1].strip
          when ""
            # Ignore
          else
            in_groonga_command = false
          end
        end
      else
        case line.chomp
        when "<!-- groonga-command -->"
          in_groonga_command = true
          in_include = false
        when /\A<!-- groonga-include\s*:/
          include_path = line.split(":", 2)[1].strip
          update_file(File.join(File.dirname(@current_path), include_path))
        end
      end
    end
    unless groonga_command_block.empty?
      process_groonga_command(groonga_command_block)
    end
  end

  def process_groonga_command(groonga_command)
    current_output_path = nil
    markup = nil
    actions = []
    command = ""
    in_load_values = false
    groonga_command.each_line do |line|
      case line
      when /\Adatabase:/
        @current_db = line.split(":", 2)[1].strip
      when /\Alog:/
        actions << {
          type: :output_log,
          value: (line.split(":", 2)[1].strip == "true"),
        }
      when /\Ainclude::/
        path = line.split("::", 2)[1].strip
        base_dir = File.dirname(@current_path)
        current_output_path = File.join(base_dir, path)
        markup = detect_markup(current_output_path)
        puts("### Current output path: #{current_output_path}")
        FileUtils.mkdir_p(File.dirname(current_output_path))
        File.open(current_output_path, "w") do |output|
          if markup == :rst
            output.puts("Execution example::")
            output.puts
          else
            output.puts("Execution example:")
            output.puts
          end
        end
      when /\A[%$] /
        actions << {
          type: :command_line,
          value: line.split(/[%$]/, 2)[1].strip,
        }
      when /\A\.\. /
        actions << {
          type: :comment,
          value: line.split("..", 2)[1].strip,
        }
      when /\A#/
        actions << {
          type: :comment,
          value: line,
        }
      else
        next if command.empty? and line == "\n"
        command << line
        next if line.end_with?("\\\n")
        if in_load_values
          in_load_values = (line != "]\n")
        else
          in_load_values = (command.start_with?("load") and
                            not command.include?(" --values "))
        end
        unless in_load_values
          actions << {
            type: :command,
            value: command,
          }
          command = ""
        end
      end
    end
    return if actions.empty?

    in_fenced_code_block = false
    run_groonga do |input, output|
      output_log = false
      actions.each do |action|
        case action[:type]
        when :output_log
          output_log = action[:value]
          puts("### Output log: #{output_log}")
        when :command_line
          command_line = action[:value]
          expanded_command_line = expand_command_line(command_line)
          puts(expanded_command_line)
          command_line_output = `#{expanded_command_line}`
          if current_output_path
            File.open(current_output_path, "a") do |output|
              if markup == :rst
                output.puts("  $ #{command_line}")
                output.puts(command_line_output.gsub(/^/, "  "))
              else
                if in_fenced_code_block
                  output.puts("```")
                  output.puts
                end
                output.puts("```console")
                output.puts("$ #{command_line}")
                output.puts(command_line_output)
                output.puts("```")
                if in_fenced_code_block
                  output.puts
                  output.puts("```shell")
                end
              end
            end
          end
          puts(command_line_output)
        when :comment
          comment = action[:value]
          if current_output_path
            File.open(current_output_path, "a") do |output|
              if markup == :rst
                output.puts(comment.gsub(/^/, "  "))
              else
                output.puts(comment)
              end
            end
          end
          puts(comment)
        when :command
          command = action[:value]
          if markup == :md and not in_fenced_code_block
            File.open(current_output_path, "a") do |output|
              output.puts("```shell")
            end
            in_fenced_code_block = true
          end
          execute_command(input, output, command, current_output_path, output_log)
        end
      end
    end
    if markup == :md and in_fenced_code_block
      File.open(current_output_path, "a") do |output|
        output.puts("```")
      end
    end
  end
end

if File.directory?("/dev/shm")
  base_dir = "/dev/shm/groonga-doc"
else
  base_dir = "/tmp/groonga-doc"
end
updator = Updator.new(base_dir)
source = ARGV[0] || ENV["SOURCES"] || File.join(__dir__, "source")
updator.update(source)
